spidermon/contrib/monitors/mixins/validation.py

```diff
@@ -84,125 +84,259 @@ def __init__(self, stats, prefix=None):
             self.validation.fields[field_name].errors["missing_required_field"].percent
         )
-        self._check_missing_required_percent(missing_percent, allowed_percent)
-
-    def _check_missing_required_percent(self, missing_percent, allowed_percent=0):
-        self.assertLessEqual(
-            missing_percent,
-            allowed_percent,
-            msg="{percent}% of required fields are missing!{threshold_info}".format(
-                percent=missing_percent * 100,
-                threshold_info=(" (maximum allowed %.0f%%)" % (allowed_percent * 100))
-                if allowed_percent > 0
-                else "",
-            ),
+        msg = self._get_msg_for_missing_required_percent(field_name, missing_percent, allowed_percent)
+        self.assertLessEqual(missing_percent, allowed_percent, msg)
+
+    @staticmethod
+    def _get_msg_for_missing_required_percent(field_name, missing_percent, allowed_percent):
+        msg = "{percent}% of required field {field} are missing!{threshold_info}".format(
+            percent=missing_percent * 100,
+            field=field_name,
+            threshold_info=(" (maximum allowed %.0f%%)" % (allowed_percent * 100))
+            if allowed_percent > 0
+            else "",
         )
+        return msg
 
     def check_fields_errors(self, field_names=None, errors=None, allowed_count=0):
-        if not field_names:
+        """
+        Checks that the number of errors for the ``field_names`` fields is less or equal than ``allowed_count`` and
+        raises an error with all problematic fields.
+        If ``field_names`` is None and ``new_behavior`` is True, checks all fields.
+        If ``field_names`` is None and ``new_behavior`` is False, checks that the total number of errors is less or
+        equal than ``allowed_count``.
+        """
+        if not self.new_behavior and not field_names:
+            # TODO: deprecate this
             errors_count = self.validation.errors.count
-            self._check_field_errors(errors_count, allowed_count)
-        else:
-            for field_name in field_names:
-                self.check_field_errors(field_name, errors, allowed_count)
+            self.assertLessEqual(
+                errors_count,
+                allowed_count,
+                msg="{count} field{count_plural} {verb} validation errors!{threshold_info}".format(
+                    count=errors_count,
+                    count_plural="" if errors_count == 1 else "s",
+                    verb="has" if errors_count == 1 else "have",
+                    threshold_info=(" (maximum allowed %d)" % allowed_count)
+                    if allowed_count > 0
+                    else "",
+                ),
+            )
+            return
+        if not field_names:
+            field_names = self._get_all_fields()
+        msgs = []
+        for field_name in field_names:
+            errors_count = self._get_errors_count(errors, field_name)
```

Comment:
I think we are missing an `if` statement here.
Otherwise, it will add a message even if there is no error for the field.
If this case holds, it would be nice to add a test case